### PR TITLE
Add system disk block resize

### PR DESCRIPTION
**Pull request by suqinhuang**
Signed-off-by: Suqin Huang <shuang@redhat.com>

id:1411207 1390485

### CODE REVIEW

File: qemu/tests/cfg/block_resize.cfg [hunk 1]
```diff
@@ -4,42 +4,57 @@
...
-        disk_rescan_cmd += " && echo exit >> cmd"
-        disk_rescan_cmd += " && diskpart /s cmd"
+    variants:
+        - with_system_disk:
+            blk_extra_params_image1 = "serial=TARGET_DISK0"
+            backup_image_before_testing = yes
+            restore_image_after_testing = yes
+            Windows:
+                block_size_cmd = wmic diskdrive get size, index
+                disk_index = 1
```
Comment (vivianQizhu) [inline qemu/tests/cfg/block_resize.cfg]: Both system and data disk index=1?

File: qemu/tests/cfg/block_resize.cfg [hunk 2]
```diff
@@ -4,42 +4,57 @@
...
-        disk_update_cmd += " && echo exit >> cmd && diskpart /s cmd"
-        disk_unit = M
-        disk_rescan_cmd = "echo rescan > cmd"
-        disk_rescan_cmd += " && echo exit >> cmd"
-        disk_rescan_cmd += " && diskpart /s cmd"
+    variants:
+        - with_system_disk:
+            blk_extra_params_image1 = "serial=TARGET_DISK0"
+            backup_image_before_testing = yes
+            restore_image_after_testing = yes
```
Comment (CongLi) [inline qemu/tests/cfg/block_resize.cfg]: Better to add 'snapshot=on' directly

File: qemu/tests/cfg/block_resize.cfg [hunk 3]
```diff
@@ -4,42 +4,57 @@
...
+        - with_data_disk:
+            images += " stg"
+            image_size_stg = 10G
+            image_name_stg = images/stg
+            create_image_stg = yes
+            force_create_image_stg = yes
+            image_snapshot_image1 = yes
+            blk_extra_params_stg = "serial=TARGET_DISK0"
+            Host_RHEL.m6..ide:
+            blk_extra_params_stg = "wwn=0x5000123456789abc"
```
Comment (CongLi) [inline qemu/tests/cfg/block_resize.cfg]: where is the indentation?

Comment (suqinhuang) [inline qemu/tests/cfg/block_resize.cfg]: pls check the original cfg file, it's inherited from the original one

File: qemu/tests/cfg/block_resize.cfg (hunk 2 above)
Comment (suqinhuang) [inline qemu/tests/cfg/block_resize.cfg]: snapshot don't support resize

File: qemu/tests/cfg/block_resize.cfg (hunk 3 above)
Comment (CongLi) [inline qemu/tests/cfg/block_resize.cfg]: From line 14, not sure is it the original cfg file which you mentioned.
    Host_RHEL.m6..ide:
         blk_extra_params_stg = "wwn=0x5000123456789abc"

**Comment by xutian** [comment]
do you mean block resize system disk?

File: qemu/tests/cfg/block_resize.cfg (hunk 3 above)
Comment (CongLi) [reply qemu/tests/cfg/block_resize.cfg]: @suqinhuang 

Could you please check it again?

Thanks.

**Comment by XueqiangWei** [comment]
passed on RHEL.7, failed on windows (Login timeout expired, waiting for xuhan's patch) 
(1/2) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.block_resize.with_system_disk: PASS (342.19 s)
 (2/2) smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.block_resize.with_data_disk: PASS (119.83 s)

@suqinhuang some comments need to be updated, thx

**Comment by peixiu** [comment]
Hi shuang,

It seems your codes don't include the disk r/w opration after disk block_resize.
According to the test case after test plan merged:
https://polarion.engineering.redhat.com/polarion/#/project/RedHatEnterpriseLinux7/workitem?id=RHEL7-94700
We also need to do some copy files actions on the extended or shrinked disk, both linux and windows.
And for windows, there are some special point need to note:
1. For windows system disk extend, after the qmp extend successfully, also need do extend the disk to the expected size on the guest, then do copy.
2. For windows system disk shrink, need do shrink the disk to the expected size on guest firstly, then do qmp shrink size action, if not, the system disk will be broken.

Best Regards~
Peixiu

File: qemu/tests/cfg/block_resize.cfg [hunk 4]
```diff
@@ -4,42 +4,57 @@
...
-        disk_rescan_cmd += " && diskpart /s cmd"
+    variants:
+        - with_system_disk:
+            blk_extra_params_image1 = "serial=TARGET_DISK0"
+            backup_image_before_testing = yes
+            restore_image_after_testing = yes
+            Windows:
+                block_size_cmd = wmic diskdrive get size, index
+                disk_index = 1
+                disk_letter = C
```
[the remark this answers is not in the document]
Comment (suqinhuang) [reply qemu/tests/cfg/block_resize.cfg]: you are right

File: qemu/tests/cfg/block_resize.cfg (hunk 3 above)
Comment (suqinhuang) [reply qemu/tests/cfg/block_resize.cfg]: if params.get("os_type") == 'linux':
        drive_id = params["blk_extra_params_%s" % data_image].split("=")[1]
        drive_path = utils_misc.get_linux_drive_path(session, drive_id)

**Comment by peixiu** [comment]
@suqinhuang Hi, I tried to test  with the latest patch on windows, block_resize.with_system_disk can be passed. but found that only run disk enlarge test, will you cover the shrink test on this patch? I checked the reference bug, need to run both shrink and enlarge test on the system disk. 

PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win8.i386.1.io-github-autotest-qemu.block_resize.with_system_disk
PASS 2-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win8.i386.1.io-github-autotest-qemu.block_resize.with_data_disk

**Comment by peixiu** [comment]
@suqinhuang Hi, I also tried to test with raw image on win8.1-64, it will failed with follows error., the enlarge system disk to 45G is successful, but the shrink disk to 15G is fail, for windows system disk shrink test, it need shrink the disk in guest device management first, then do shrink on qmp, hope that's helpfull.

06:35:18 ERROR|
06:35:18 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-53.0-py2.7.egg/avocado_vt/test.py:245
06:35:18 ERROR| Traceback (most recent call last):
06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-53.0-py2.7.egg/avocado_vt/test.py", line 397, in _runTest
06:35:18 ERROR|     run_func(self, params, env)
06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
06:35:18 ERROR|     return fn(*args, **kwargs)
06:35:18 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/block_resize.py", line 149, in run
06:35:18 ERROR|     session.cmd(disk_rescan_cmd)
06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1238, in cmd
06:35:18 ERROR|     print_func)
06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1178, in cmd_status_output
06:35:18 ERROR|     raise ShellStatusError(cmd, o)
06:35:18 ERROR| ShellStatusError: Could not get exit status of command: 'echo rescan > cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.3.9600\n\nCopyright (C) 1999-2013 Microsoft Corporation.\nOn computer: WIN-ACNHAKD8LMU\n\nDiskPart was unable to open or read the script file.\nMake sure the file you specified exists.\n\n')
06:35:18 ERROR|
06:35:18 ERROR|
06:35:18 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_framework-53.0-py2.7.egg/avocado/core/test.py:630
06:35:18 ERROR| Traceback (most recent call last):
06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-53.0-py2.7.egg/avocado_vt/test.py", line 271, in runTest
06:35:18 ERROR|     raise self.__status  # pylint: disable=E0702
06:35:18 ERROR| ShellStatusError: Could not get exit status of command: 'echo rescan > cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.3.9600\n\nCopyright (C) 1999-2013 Microsoft Corporation.\nOn computer: WIN-ACNHAKD8LMU\n\nDiskPart was unable to open or read the script file.\nMake sure the file you specified exists.\n\n')
06:35:18 ERROR|
06:35:18 DEBUG| Local variables:
06:35:18 DEBUG|  -> self <class 'avocado_vt.test.VirtTest'>: 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win8.x86_64.1.io-github-autotest-qemu.block_resize.with_system_disk
06:35:18 ERROR| Traceback (most recent call last):

06:35:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_framework-53.0-py2.7.egg/avocado/core/test.py", line 701, in _run_avocado
    raise test_exception

06:35:18 ERROR| ShellStatusError: Could not get exit status of command: 'echo rescan > cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.3.9600\n\nCopyright (C) 1999-2013 Microsoft Corporation.\nOn computer: WIN-ACNHAKD8LMU\n\nDiskPart was unable to open or read the script file.\nMake sure the file you specified exists.\n\n')

06:35:18 ERROR| ERROR 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win8.x86_64.1.io-github-autotest-qemu.block_resize.with_system_disk -> ShellStatusError: Could not get exit status of command: 'echo rescan > cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.3.9600\n\nCopyright (C) 1999-2013 Microsoft Corporation.\nOn computer: WIN-ACNHAKD8LMU\n\nDiskPart was unable to open or read the script file.\nMake sure the file you specified exists.\n\n')

**Comment by CongLi** [comment]
@peixiu @XueqiangWei 

Could you please help review this PR now?

**Comment by peixiu** [comment]
Tried to test this case with with raw image on Win7-32, with_system_disk block resize tests also failed. Test result and error message as follows:

 (2/5) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.with_system_disk: ERROR (1188.79 s)
 (3/5) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.with_data_disk: PASS (114.60 s)
 (4/5) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.data_plane.with_system_disk: ERROR (815.48 s)
 (5/5) smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.data_plane.with_data_disk: PASS (122.96 s)

06:56:10 ERROR|
06:56:10 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-54.0-py2.7.egg/avocado_vt/test.py:245
06:56:10 ERROR| Traceback (most recent call last):
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-54.0-py2.7.egg/avocado_vt/test.py", line 397, in _runTest
06:56:10 ERROR|     run_func(self, params, env)
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
06:56:10 ERROR|     return fn(*args, **kwargs)
06:56:10 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/block_resize.py", line 145, in run
06:56:10 ERROR|     session.cmd(disk_update_cmd[index])
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1238, in cmd
06:56:10 ERROR|     print_func)
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1165, in cmd_status_output
06:56:10 ERROR|     o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1086, in cmd_output
06:56:10 ERROR|     raise ShellTimeoutError(cmd, o)
06:56:10 ERROR| ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'echo select disk 0 > cmd && echo select partition 2 >> cmd && echo shrink desired=30720 >> cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.1.7601\nCopyright (C) 1999-2008 Microsoft Corporation.\nOn computer: WIN-P67DHK0VAON\n\nDisk 0 is now the selected disk.\n\nPartition 2 is now the selected partition.\n')
06:56:10 ERROR|
06:56:10 ERROR|
06:56:10 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_framework-54.1-py2.7.egg/avocado/core/test.py:664
06:56:10 ERROR| Traceback (most recent call last):
06:56:10 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-54.0-py2.7.egg/avocado_vt/test.py", line 271, in runTest
06:56:10 ERROR|     raise self.__status  # pylint: disable=E0702
06:56:10 ERROR| ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'echo select disk 0 > cmd && echo select partition 2 >> cmd && echo shrink desired=30720 >> cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.1.7601\nCopyright (C) 1999-2008 Microsoft Corporation.\nOn computer: WIN-P67DHK0VAON\n\nDisk 0 is now the selected disk.\n\nPartition 2 is now the selected partition.\n')
06:56:10 ERROR|
06:56:10 DEBUG| Local variables:
06:56:11 DEBUG|  -> self <class 'avocado_vt.test.VirtTest'>: 2-smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.with_system_disk
06:56:11 ERROR| Traceback (most recent call last):

06:56:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_framework-54.1-py2.7.egg/avocado/core/test.py", line 735, in _run_avocado
    raise test_exception

06:56:11 ERROR| ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'echo select disk 0 > cmd && echo select partition 2 >> cmd && echo shrink desired=30720 >> cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.1.7601\nCopyright (C) 1999-2008 Microsoft Corporation.\nOn computer: WIN-P67DHK0VAON\n\nDisk 0 is now the selected disk.\n\nPartition 2 is now the selected partition.\n')

06:56:11 ERROR| ERROR 2-smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.raw.virtio_blk.up.virtio_net.Win7.i386.sp1.io-github-autotest-qemu.block_resize.with_system_disk -> ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'echo select disk 0 > cmd && echo select partition 2 >> cmd && echo shrink desired=30720 >> cmd && echo exit >> cmd && diskpart /s cmd'    (output: u'\nMicrosoft DiskPart version 6.1.7601\nCopyright (C) 1999-2008 Microsoft Corporation.\nOn computer: WIN-P67DHK0VAON\n\nDisk 0 is now the selected disk.\n\nPartition 2 is now the selected partition.\n')
06:56:11 INFO |

**Comment by peixiu** [comment]
Closed this patch, it is re-implemented with follows patch, fixed the issue mentioned on my pevious comment.
https://github.com/autotest/tp-qemu/pull/1266
https://github.com/avocado-framework/avocado-vt/pull/1454
https://github.com/avocado-framework/avocado-vt/pull/1455

**Comment by luckyh** [comment]
Close per the last comment.

**Pull request closed without merging**